Feature: (optional) UM can ignore identity changes emitted from stream

JordanMartinez opened the pull request:
- also adds javadoc to UndoManagerFactory and UndoManager
- also uses default methods to reduce code duplication in the factory

Address #14

JordanMartinez commented:
Since there's no Travis CI on here:
````
~/Git Projects/UndoFX $ git branch
* addIdentity
  master
  placeholderForRound2Work
  undo-redo
~/Git Projects/UndoFX $ gradle test
:undofx:compileJava
:undofx:processResources UP-TO-DATE
:undofx:classes
:undofx:compileTestJava
:undofx:processTestResources UP-TO-DATE
:undofx:testClasses
:undofx:test
:undofx:jar
:undofx-demos:compileJava
:undofx-demos:processResources UP-TO-DATE
:undofx-demos:classes
:undofx-demos:compileTestJava UP-TO-DATE
:undofx-demos:processTestResources UP-TO-DATE
:undofx-demos:testClasses UP-TO-DATE
:undofx-demos:test UP-TO-DATE

BUILD SUCCESSFUL

Total time: 1.455 secs
````

JordanMartinez commented:
Actually, let me update the build file to resolve links to ReactFX properly, too.

JordanMartinez commented:
Ok. Good to go.

TomasMikula commented:
You forgot the main thing: test for identity after merging. This can be done in the `private C[] merge(C c1, C c2)` method, returning an empty array when the result of `merge` is identity.

Also, to reduce the number of constructors, a constructor should take `isIdentity` only if it also takes `merge`. When there is no `merge`, instead of  passing in `isIdentity`, the user can just filter `changeStream` with `!isIdentity`.

JordanMartinez commented:
> You forgot the main thing: test for identity after merging. This can be done in the private C[] merge(C c1, C c2) method, returning an empty array when the result of merge is identity.

Sorry. Somewhere in the process, I thought the `isIdentity` only applied to the changes that are emitted by the eventstream, not the merge process. Re-reading through #14, I realize that's not the case. I'll fix it possibly tomorrow but definitely by next week.

JordanMartinez commented:
Fixed.

On another thought, should there be a 3rd constructor with a `merge` parameter but no `isIdentity` parameter because it assumes that every emitted event and merged event is non-identity?

TomasMikula commented:
Yes. That also preserves compatibility with the previous version.

JordanMartinez commented:
Fixed

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java:
```diff
@@ -209,7 +229,7 @@ private void addChange(C change) {
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
             return (C[]) new Object[] { merged.get() };
         } else {
             return (C[]) new Object[] { c1, c2 };
```
Why not return an empty array when merge results in an identity change?

JordanMartinez replied:
When the change stream emits a non-identity change, `addChange` is called, right? (I've in-lined the `merge` method below:
````java
@SuppressWarnings("unchecked")
private void addChange(C change) {
    if(canMerge && queue.hasPrev()) {
        C prev = queue.prev();
        // merge in-lined start
        Optional<C> merged = merge.apply(prev, change);
        C[] changeArray;
        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
            changeArray = (C[]) new Object[] { merged.get() };
        } else {
            changeArray = (C[]) new Object[] { c1, c2 };
        }
        // merge in-lined end
        queue.push(changeArray);
    } else {
        queue.push(change);
    }
    canMerge = true;
    undoAvailable.invalidate();
    redoAvailable.invalidate();
    atMarkedPosition.invalidate();
}
````
If I return an empty array, won't that mean the emitted change is not stored at all?

TomasMikula replied:
Yes, that's what I intended.

Taking RichTextFX as an example, say you type in `hello` and then hit backspace 5 times. With my version, the undo history will now be empty. With yours, it will contain 2 items:
 - at 0, insert `h`
 - at 0, delete `h`

I would say yours is as far from what actually happened as mine.

Plus, why merge the  first 4 deletions and not the 5th one? Seems somewhat irregular and arbitrary.

Plus, if the user didn't want changes to annihilate, they could have provided `merge` that would return `Optional.empty()` instead of an identity change. Thus this behavior could be implemented without any new support from UndoFX.

TomasMikula replied:
Now that you inlined it, I see one problem, though. If the changes can annihilate, it can cause merging of changes that were previously prevented from being merged.

When this is the current queue

```
... c1
```

and `preventMerge` is invoked, there will be a _temporary_ barrier for merging `c1` (`canMerge` set to `false`):

```
... c1|
```

Then when `c2` arrives, merge will not even be attempted, but the barrier is erased.

```
... c1 c2
```

Then if `c3` = <code>c2<sup>-1</sup></code> is received, `c2` is annihilated and `c1` is no longer prevented from being merged.

```
... c1
```

JordanMartinez replied:
> Yes, that's what I intended.
> 
> Taking RichTextFX as an example, say you type in hello and then hit backspace 5 times. With my version, the undo history will now be empty. With yours, it will contain 2 items:
> 
>     at 0, insert `h`
>     at 0, delete `h`
>
> I would say yours is as far from what actually happened as mine.
> 
> Plus, why merge the first 4 deletions and not the 5th one? Seems somewhat irregular and arbitrary.
>
> Plus, if the user didn't want changes to annihilate, they could have provided merge that would return Optional.empty() instead of an identity change. Thus this behavior could be implemented without any new support from UndoFX.

Took me a while to understand what you meant. Rephrasing it in my own words, if a user typed in "hello", this would result in five changes being emitted by the change stream. When `merge` is called, these are merged into one change `insert: "hello" | removed: ""`. When the user deletes the last four letters, the `merge` will gradually remove the letters in the `insert` part of the change: `hello -> hell -> hel -> he -> h`. On the final deletion, it would attempt to merge `insert: "h" | removed: ""` with `insert: "" | removed: "h"`. The result would be an identity change `inserted: "" | removed: ""`. 

There are two ways to handle this situation. The first is what my code does: "the merge was successful, but it's an identity change, so just re-store the previous change and store the new pushed change". As a result, the queue would have two changes. The second is what you have said should occur: "the merge was successful, but it's an identity change, so act as though neither the stored change nor the pushed change were ever done in the first place." In effect, the changes have been "annihilated." 
If my approach is used and one doesn't want "change annihilation" in their code, there is nothing they can do about it. If your approach is used and they don't want it, they can simply modify the `merge` to return `Optional.empty()` in the situation where a successful merge would return an identity change.
Thus, your approach is more flexible between the two.

JordanMartinez replied:
> Now that you inlined it, I see one problem, though. If the changes can annihilate, it can cause merging of changes that were previously prevented from being merged.

To rephrase this in my own words...
Given a change that is prevented from being merged with another change (unmergeable change), a change annihilation that occurs directly after the unmergeable change should not allow the unmergeable change to be merged with another pushed change. This implies that `canMerge` will need to be set differently in different circumstances.

TomasMikula replied:
A single boolean flag (`canMerge`) will not suffice, since we can no longer forget that the user called `preventMerge` once the next change (`c2` in my previous comment) arrives. One way to store in the queue `Either<Barrier, C>`, where left case means to prevent merge of changes to the left and right of the "barrier". (`Barrier` could actually be `Void`, or, if we insist that user's changes are not `null`, we could just use misuse `null` to mean the barrier, to avoid too much boxing.

JordanMartinez replied:
> A single boolean flag (canMerge) will not suffice, since we can no longer forget that the user called preventMerge once the next change (c2 in my previous comment) arrives. 

Mm... true. I assumed the change annihilation would occur every time `preventMerge` is called, but that doesn't happen every time.

> One way to store in the queue Either<Barrier, C>, where left case means to prevent merge of changes to the left and right of the "barrier".

Not sure where/how `Either<Barrier, C>` would be used. I need a bit more context to connect the dots here.

One question I have: when a user calls `preventMerge`, is the previous change in the queue (the next undo) supposed to never be merged with any future changes? Given the following timeline where `|` represents the temporary barrier and `~` represents the current position in the queue:
````
.... c1 |~ // prevent merge called
.... c1 c2 c3 ~ // more changes pushed
.... c1 ~ c2 c3 // undo called twice
````
When a new change (`c4`) is emitted, should `c1` merge with `c4`? It would seem the latest release would merge them. Is that supposed to happen?

If not, perhaps we can make the following changes:
- The type in `List<C> changes` becomes `List<Either<C, C>> changes`
- the left side would represent an unmergeable change
- the right side would represent a mergeable change
- when `preventMerge` is called, we'd call 
````
if (currentPosition > 0) {
    Either<C, C> either = changes.get(currentPosition - 1);
    if (either.isRight()) {
        C c = either.getRight();
        changes.set(currentPosition - 1, Either.left(c));
        nextIsUnmergeable = false;
    }
} else {
    nextIsUnmergeable = true;
}
````
- when a change is pushed and `nextIsUnmergeable` is true, it stores the change as left and not as right.

TomasMikula replied:
> Not sure where/how `Either<Barrier, C>` would be used.

I just meant that the type of queue would change from 

```java
ChangeQueue<C> queue;
```

to

```java
ChangeQueue<Either<Barrier, C>> queue;
```

Seems like you are proposing to change it to

```java
ChangeQueue<Either<C, C>> queue;
```

which would allow us to preserve the information about which changes are mergeable as well.

Neither of these might be necessary for now, after all. (See below.)

> ````
> .... c1 |~ // prevent merge called
> .... c1 c2 c3 ~ // more changes pushed
> .... c1 ~ c2 c3 // undo called twice
> ````
> When a new change (`c4`) is emitted, should `c1` merge with `c4`? It would seem the latest release would merge them. Is that supposed to happen?

It wouldn't, thanks to [this line](https://github.com/TomasMikula/UndoFX/blob/3d5555cfd84e3740e4cee004ada51abeebb93789/undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java#L93) that sets `canMerge` to `false`. It is questionable whether this the right thing to do (one might want to prevent merge at this point only if `preventMerge` was called after `c1`, but that information is now lost), but no one has complained yet :)

If we, for now, allow ourselves the same amount sloppiness, we could also just set `canMerge` to `false` after annihilation, and do "the right thing" in a separate PR.

JordanMartinez replied:
One thought I had about the `Either<C, C>` approach. When I implement the non-linear PR, wouldn't it make sense to have a `preventBubble` method? If so, how should that meta-data be stored? Then a change could be

mergeable | bubble-able
--------------|--------------
x | x
x | _
_ | x
_ | _

If we follow the `Either<Barrier, C>` approach, `Barrier` could store all three: the change itself, whether it can be merged, and whether it can be bubbled.

JordanMartinez commented:
I've updated it to use the empty array approach now.

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java:
```diff
@@ -209,8 +229,10 @@ private void addChange(C change) {
-        if(merged.isPresent()) {
-            return (C[]) new Object[] { merged.get() };
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
+            return isNotIdentity.test(merged.get())
```
Something's not right: the second `isNotIdentity.test` always returns `true`.

JordanMartinez replied:
WOW!.... Can't believe I forgot to remove the first test and only have the second... Sheez...

JordanMartinez replied:
I've fixed this now on the PR. However, I haven't addressed the sloppy `canMerge` part as well on my local version because the test I wrote for it failed. I'll have to try later when I have more time.

JordanMartinez replied:
K. I've fixed it now.
I in-lined `merge` into the method because it makes it easier to read since one has to set `canMerge` on each part.
I also split up the tests so that one doesn't interfere with another.

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/UndoManagerFactory.java:
```diff
@@ -14,118 +15,249 @@
+     *
+     * @param changeStream the {@link EventStream} that emits changes
+     * @param invert inverts the change
+     * @param apply applies the change (undo/redo)
```
Given the names and type signatures of these last three parameters, do these descriptions add any information? I don't think so.

JordanMartinez replied:
No, they don't. I guess I was going for completeness.

TomasMikula replied:
But there is useful information to add that is not captured by the type signatures.

What does it mean to do `apply(invert(c))`?

What should be expected to be emitted from `changeStream` after a call to `apply(c)`?

In the presence of `isIdentity`, what property should hold for `merge(c, invert(c))`?

JordanMartinez replied:
Can you document that then? Otherwise, I feel like this PR will take much longer because I'll write something down and then you'll point out something that needs to be changed.

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/UndoManagerFactory.java:
```diff
@@ -14,118 +15,249 @@
+     * @param invert inverts the change
+     * @param apply applies the change (undo/redo)
+     * @param merge merges two changes into one change. If the resulting change is an identity change, the two
+     *              changes will not be merged.
```
This comment is already outdated and lying. This is the trouble with stringly typed documentation: gets out of sync with code so easily.

JordanMartinez replied:
True! Currently fixing that.

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/UndoManagerFactory.java:
```diff
@@ -14,118 +15,249 @@
+     * Creates an {@link UndoManager} with no history where all changes emitted from the EventStream will be
+     * considered non-identity changes and each emitted change will change the
+     * {@link ChangeQueue#getCurrentPosition()}; the {@link UndoManager} will never be able to undo/redo a change
+     * emitted from the changeStream.
```
For the user, `ChangeQueue` is an implementation detail. Thus the documentation of the public API should not mention it.

JordanMartinez replied:
True. I added that because I wasn't sure how else to explain that `atMarkedPosition` wouldn't be the same value after a push. I guess I could just say that instead.

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java:
```diff
@@ -68,16 +70,32 @@ protected boolean computeValue() {
         this.invert = invert;
         this.apply = apply;
         this.merge = merge;
+        this.isNotIdentity = isIdentity.negate();
```
Negative conditions are harder to read. I would keep it as `isIdentity`.

JordanMartinez replied:
Yeah, I was wondering about that. Sometimes, I think it's easier to read a `notCondition` than a `!condition` because the latter's mark is easier to miss.

JordanMartinez commented:
I've addressed your comments.

Additionally, I removed the javadoc for `merge` when `isIdentity = c -> false` because there's no point in mentioning the change annihilation part since that will never happen in those cases.

TomasMikula commented on undofx/src/main/java/org/fxmisc/undo/UndoManagerFactory.java:
```diff
@@ -14,118 +15,215 @@
+     * Creates an {@link UndoManager}.
+     *
+     * @param merge merges the next undo and the most recently pushed change. If the resulting change is an identity
+     *              change, neither of the changes will be stored (even the next undo).
```
Document what does it mean for `merge` to return empty `Optional`?

JordanMartinez replied:
Can you document this, too? (see comment in other review)

JordanMartinez commented:
Since I haven't heard anything back from you about the documentation issue, I'd like to clarify some things to insure no offense was made and to insure this gets done. I initially removed the javadoc concerning the `changeStream`, `apply`, and `invert` parameters because I thought your comment implied it was pointless to document them. Then you later implied that documenting these parameters  was worth it, and listed what could be documented. You also indicated another place where documentation could be improved `merge`. Since this PR was taking longer than expected (I take blame for misunderstanding and misreading), taking time away from papers I still need to write before my semester ends, and since you implied that additional documentation was needed but did not specify what exactly, so that I could write it quickly and easily, I thought that you should document it because it would meet your standards and be immediately merged whereas it might be delayed by miscommunication between us via comments of "this could be better documented" or "you forgot to do this." So, while I'm guessing you haven't responded due to being busy with other matters and haven't been offended, I want to insure I did not somehow offend you by my comment.

With that being said, I'd also guess that it would be inconvenient for you to literally document this yourself: you would need to add my repo, check out a local copy, document everything, submit the changes to my repo, where I would then have to merge it, and then you would merge this PR here. Thus, it would be better if I could fix the documentation and update this PR. Since I am unsure what I should write, I think it would be faster if you wrote the documentation via comments and I updated my PR to include those comments in the documentation. Your thoughts?

TomasMikula commented:
Don't worry, no offense taken, just being busy ;)

> I initially removed the javadoc concerning the `changeStream`, `apply`, and `invert` parameters because I thought your comment implied it was pointless to document them.

I only meant to imply that those specific comments were pointless (duplicating the code).

> you implied that additional documentation was needed

You said you were going for completeness. I pointed out there was something worthwhile to document [if you really wanted to].

> but did not specify what exactly, so that I could write it quickly and easily

I just wanted to give you the chance to figure it out yourself (if you wanted). If not, I can as well just merge (since I'm already happy with the code) and update comments myself afterwards.

TomasMikula commented:
Good luck with your papers 😉

JordanMartinez commented:
Ah, thanks for the clarification. That makes sense now.

> I just wanted to give you the chance to figure it out yourself (if you wanted). If not, I can as well just merge (since I'm already happy with the code) and update comments myself afterwards.

Thanks for not taking away my opportunity to learn. While I think completeness is best, I probably won't be able to get to this until after my semester finishes (mid-may ish). So, the goal is to fix RTFX ASAP so that a new release without any regressions is made, then it makes sense to merge this now. If you're willing to wait, then I could do it after my semester finishes, but the context of our conversation will not be as fresh in my mind.

> Good luck with your papers

Thank you sir!

JordanMartinez commented:
Just fetched your new updates. Why did you add the `nextToUndo/Redo`? (And wouldn't `nextUndo/Redo` be easier to read than what you have?)

TomasMikula commented:
Yeah, I guess maybe those would have been better names...

The motivation is to be able to peek at the next/previous change, so that one can display
"Undo: Insert text", "Redo: Move object" as menu item labels or button tooltips.

JordanMartinez commented:
> The motivation is to be able to peek at the next/previous change, so that one can display
"Undo: Insert text", "Redo: Move object" as menu item labels or button tooltips.

Mm.. Good idea.

> Yeah, I guess maybe those would have been better names...

Maybe in UndoFX 1.3.2?

TomasMikula commented:
Yeah, but I'm not going to make another release just for that.

JordanMartinez commented:
Makes sense.
